dxr/plugins/clang/needles.py

```diff
@@ -35,24 +30,50 @@ def needles(condensed, name, suffix='', kind=None, subkind=None, keys=('name', '
          ``condensed``. Defaults to the value of ``name``.
     :arg subkind: The value of the 'kind' key to insist on in things found
         within ``condensed[kind]``. None for no insistence.
-    :arg keys: The keys that should be in the final needle value
+    :arg include_qualname: Whether we should include the qualified name in the
+        returned needle value dict
+    :arg include_typeless_qualname: Whether we should include the qualname but
+        with arg types stripped off, e.g. turning functions from Foo::bar(int)
+        to Foo::bar, so people don't need to know the types to search for
+        them. ``include_qualname`` must also be true to enable this.
 
     """
     kind = kind or name
     matches_subkind = (lambda entity: entity.get('kind') == subkind if subkind
                        else lambda entity: True)
+
+    def names(entity):
+        """Return the name, qualname, and type-free qualname of an entity.
+
+        This way, Moo::foo(int) can be found as "foo", "Moo::foo" or
+        "Moo:foo(int)".
+
+        """
+        value = {'name': entity['name']}
+        if include_qualname:
+            qualname = value['qualname'] = entity['qualname']
+            if include_typeless_qualname:
+                try:
+                    # This is good enough for functions but probably not for
+                    # templates yet:
+                    offset = qualname.index('(')
+                except ValueError:
+                    pass
+                else:
+                    value['qualname'] = [qualname, qualname[:offset]]
+        return value
+
     return (('c_{0}{1}'.format(name, suffix),
-             dict((k, entity[k]) for k in keys),
+             names(entity),
              entity['span'])
             for entity in condensed[kind] if matches_subkind(entity))
 
 
-def qualified_needles(condensed, name, kind=None):
-    """Return needles for a top-level kind of thing that has a name and qualname."""
-    return needles(condensed, name, kind=kind)
-
-
-def ref_needles(condensed, name, subkind=None, keys=('name', 'qualname')):
+def ref_needles(condensed,
+                name,
+                subkind=None,
+                include_qualname=True,
+                include_typeless_qualname=False):
     """Return needles for references to a certain kind of thing.
```

Comment:
I realize you didn't change the line in this request, but I'm reading "certain kind of thing" and finding myself asking "what kind of what thing?" Can we clarify this?